feat: move new client script

JoeWang1127 opened the pull request:
In this PR:
- Migrate `new-client.py` from google-cloud-java, because we will be using new-client.py to other use cases than google-cloud-java repository, such as self-service client library generation.
- Separate repo-level process script to `library_generation/repo-level-postprocess`

To verify `new-client.py`:
1. pull the branch.
2. pull the main branch of `google-cloud-java`.
3. copy `library_generation` folder to root directory of `google-cloud-java`.
4. in `google-cloud-java`, remove a newly added module, e.g., `java-cloudquotas` ([PR](https://github.com/googleapis/google-cloud-java/pull/10136)).
5. regenerate the module:
```
python library_generation/new_client/new-client.py generate --api_shortname=cloudquotas --proto-path=google/api/cloudquotas --name-pretty="Cloud Quotas API" --product-docs="https://cloud.google.com/cloudquotas/docs/" --api-description="Cloud Quotas API provides GCP service consumers with management and
    observability for resource usage, quotas, and restrictions of the services
    they consume."
```
6. use `git status` and `git diff` to verify the difference.

This is part of [milestone 2](https://docs.google.com/document/d/1v-sJBmdNEVBRryL8n90XK8OIiqVphaJemXVhANw14Jk/edit?resourcekey=0-QGUb2do-JBlDWKvptWBp_g&tab=t.0#bookmark=id.ejmjovgdfgrf) of hermetic build project plan.

Note that I didn't test the script against split repositories because the main use case of this script is to generate a module in google-cloud-java.

JoeWang1127 commented on library_generation/new_client/new-client.py:
```diff
@@ -0,0 +1,468 @@
+    if api_shortname == "":
+        sys.exit("api_shortname is empty")
+
+    repo = "googleapis/google-cloud-java"
```
This is the right value since we only use the script to generate library inside the monorepo.

diegomarquezp commented on library_generation/new_client/client_inputs.py:
```diff
@@ -0,0 +1,146 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
At a glance, it looks like this script parses BUILD files into arguments (for generate_library?). Can you please clarify in a comment?

JoeWang1127 replied:
Added.

diegomarquezp commented on library_generation/new_client/new-client.py:
```diff
@@ -0,0 +1,458 @@
+    if rpc_docs:
+        repo_metadata["rpc_documentation"] = rpc_docs
+    # Initialize workdir
+    workdir = Path(f"{sys.path[0]}/../../output/java-{api_shortname}").resolve()
```
Is this the `output` folder created by the other `library_generation` scripts? I think it may change depending on `cwd` (`../../output` may not work) if that's the case

diegomarquezp replied:
maybe we can `def` a function that computes the path, just like [`get_output_folder`](https://github.com/googleapis/sdk-platform-java/blob/265b446af6b66c3798917bc2c489474c5a482173/library_generation/utilities.sh#L212-L218). We can then move it to utils.py once the composed_library PR is merged.

JoeWang1127 replied:
> Is this the output folder created by the other library_generation scripts?

No, this is the default output folder in `generate_library.sh`.

Since new-client.py controls how to call `generate_library.sh`, I think it's safe to assume `generate_library.sh` will generate/download everything into `output` folder.

diegomarquezp replied:
> default output folder
On a second thought, does this mean that `output` will always be created at the root of our local `sdk-platform-java` folder?

JoeWang1127 replied:
> On a second thought, does this mean that output will always be created at the root of our local sdk-platform-java folder?

Yes.

diegomarquezp replied:
I see. We may have a problem when runnig the scripts somewhere other than a `/path/to/local/sdk-platform-java`. The scripts are currently written to [generate the `output` folder wherever the user is calling the scripts from](https://github.com/googleapis/sdk-platform-java/blob/265b446af6b66c3798917bc2c489474c5a482173/library_generation/utilities.sh#L212-L218) (e.g. if I call from `sdk-platform-java/showcase/scripts` then a `sdk-platform-java/showcase/scripts/output` folder will be created.

This may restrict the user to only call `new-client.py` from `sdk-platform-java`'s root. I would lean for `new-client.py` to use the value from `utilities.sh -> get_output_folder`

diegomarquezp commented on library_generation/new_client/new-client.py:
```diff
@@ -0,0 +1,458 @@
+        "--gapic_generator_version",
+        generator_version,
+        "--protobuf_version",
+        "23.2",
```
I think we can use [`sh_util`](https://github.com/googleapis/sdk-platform-java/blob/265b446af6b66c3798917bc2c489474c5a482173/library_generation/utilities.py#L50) from the composed library PR to get the protobuf version

diegomarquezp commented on library_generation/new_client/new-client.py:
```diff
@@ -0,0 +1,458 @@
+# Copyright 2019 Google LLC
```
Since we have a few modifications here and there, I think it makes sense to update the copyright year

diegomarquezp commented on library_generation/new_client/templates/owlbot.py.j2:
```diff
@@ -0,0 +1,24 @@
+    s.move(library)
+
+s.remove_staging_dirs()
+{% if should_include_templates %}java.common_templates(monorepo=True, {% if template_excludes %}excludes={{ template_excludes }}{% endif %}){% endif %}
```
That `template_excludes` variable always comes as an empty array when called from `new-client.py`, so maybe it's better to just hardcode it here. I guess this happens because this template file was recycled from somewhere else

JoeWang1127 replied:
I think maybe it's good to keep it as-is because we don't want to change the logic in this PR.

We can make refactor in further PRs.

JoeWang1127 commented on library_generation/new_client/new-client.py:
```diff
@@ -0,0 +1,458 @@
+        sys.exit(
+            "Couldn't create the module because "
+            f"the module {workdir} already exists. In Java client library "
+            "generation, a new API version of an existing module does not "
```
> does "an existing module does not require new-client.py invocation" still hold for our new scripts?

Do you mean the script that parse the config file and call `generate_composed_library.py`?

If yes, then the answer is no. I think the script can do at least two things:
- generate a new library (given an new entry or a new gapic in config)
- regenerate an existing library.

diegomarquezp commented on library_generation/owlbot/.gitignore:
```diff
@@ -0,0 +1 @@
+google-java-format.jar
```
@blakeli0 FYI we ended up reverting the use of `fmt:format` because the format plugin is inherited from `google-cloud-jar-parent`, which is fixed _after_ postprocessing is run.

blakeli0 replied:
Can you please elaborate? Why can't we use `mvn fmt:format` in the end after the poms are generated?

diegomarquezp replied:
Using `fmt` gives a missing versions [error](https://paste.googleplex.com/6024683240030208)

That's because these versions come from a parent pom not yet fixed (the fix occurs [here](https://github.com/googleapis/sdk-platform-java/blob/d2367889d753ddc25185aa2e47cb40be1ebb2a4f/library_generation/new_client/new-client.py#L379-L386), after postprocessing).

JoeWang1127 replied:
One way to fix this is the change the template; or move the format to the last step.

JoeWang1127 replied:
This jar file is removed.

suztomo commented on library_generation/repo-level-postprocess/readme_update.sh:
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+# This script should run as a part of new-library generation process.
```
I expect README.md to be generated from template.

JoeWang1127 replied:
> I expect README.md to be generated from template.

I agree that this should be done, but should we do it in this PR since the goal of this PR is migration?

suztomo replied:
Would you elaborate why this migration needs to change how README.md is generated?

JoeWang1127 replied:
As discussed in chat, this is how README.md modified by new-client.py in google-cloud-java.

JoeWang1127 replied:
Try to change the template of README.md to remove this script.

JoeWang1127 replied:
The template of `README.md` has release-please annotation and this script is no longer needed.

suztomo commented on library_generation/repo-level-postprocess/delete_non_generated_samples.sh:
```diff
@@ -0,0 +1,16 @@
+
+set -e
+
+while IFS= read -r samples_dir; do
```
Can we now stop the root cause of generating unnecessary samples?

JoeWang1127 replied:
I think it related to pom templates in [synthtool](https://github.com/googleapis/synthtool/tree/master/synthtool/gcp/templates/java_library/samples).

@diegomarquezp Do you depend on pom.xml in this directory? Can I remove them?

diegomarquezp replied:
@JoeWang1127 No, I think the poms related to our work are in the [owlbot implementation](https://github.com/googleapis/sdk-platform-java/tree/main/library_generation/owlbot/templates/poms)

suztomo replied:
@JoeWang1127 Did you get the answer how to stop generating unnecessary samples into monorepo?

JoeWang1127 replied:
I think this issue can be solved by hacking into synthtool, for example change [_generic_library](https://github.com/googleapis/synthtool/blob/af16e6d4672cc7b400f144de2fc3068b54ff47d2/synthtool/gcp/common.py#L52).

However, should we do it now or after migrating all java specific functions to sdk-platform-java? @diegomarquezp what do you think of this problem?

diegomarquezp replied:
@JoeWang1127 We have a pending cleanup consisting of moving all the templates in `synthtool/synthtool` to `sdk-platform-java/library_generation/owlbot`. We initially meant to "move" the templates but we can also get rid of a few that are not used, such as `CONTRIBUTING.md` or the samples mentioned here. 

I would say this unnecessary-sample-fix should be done in that cleanup PR

JoeWang1127 replied:
@suztomo Do you think it's reasonable to defer the refactor to further PRs?

suztomo replied:
That's sounds good. Avoid copying `delete_non_generated_samples.sh`.

JoeWang1127 replied:
We'll use the hermetic build script to generate google-cloud-java, so the current approach is to copy the script and refactor in later PR.

JoeWang1127 replied:
Since I changed the template of `owlbot.py`, we don't need this script anymore.

suztomo commented on library_generation/repo-level-postprocess/apply_current_versions.sh:
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+# This script sets the "current-version" written in versions.txt applied to all
```
This script is available in google-cloud-java. When you generate a new client library there, you have access to this script.

JoeWang1127 replied:
I plan to migrate these files as well since they can be used in repo-level post processing.

suztomo replied:
> in repo-level post processing.

Does this include other repositories than google-cloud-java? If so that makes sense.

JoeWang1127 replied:
> Does this include other repositories than google-cloud-java? If so that makes sense.

Yes, these scripts (in `repo-level-postprocess`) are applied to all java modules in google-cloud-java.

suztomo replied:
apply_current_versions.sh is not needed any more?

suztomo replied:
> Does this include other repositories than google-cloud-java?
>  all java modules in google-cloud-java

If the scope is within google-cloud-java, there's no point to copy this to sdk-platform-java.

JoeWang1127 replied:
This script is no longer needed.

blakeli0 commented on library_generation/owlbot/templates/poms/cloud_pom.xml.j2:
```diff
@@ -62,6 +62,10 @@
       <groupId>com.google.api</groupId>
       <artifactId>gax-httpjson</artifactId>
     </dependency>
+    <dependency>
```
Is this dependency a must have for most of the new modules? If yes, I think we can make it a `test` as grpc modules are only used by unit tests. See examples in [google-cloud-java](https://github.com/googleapis/google-cloud-java/blob/4982e78e416fec80baff99f0131815e2d4955acf/java-speech/google-cloud-speech/pom.xml#L108-L112)

JoeWang1127 replied:
I think the test scope is manually added ([commit](https://github.com/googleapis/google-cloud-java/pull/9258/commits/7512e3a0a0221070f8cf7991775033267d80dac1)).

We can make this change through template but I suggest we do it after the migration, along with other improvements.

suztomo replied:
I see it's been there https://github.com/googleapis/synthtool/blob/master/docker/owlbot/java/templates/poms/cloud_pom.xml.j2#L67 for few years in synthtool's template.

JoeWang1127 replied:
Thanks @suztomo. 

@diegomarquezp do you remember deleting this dependency when you migrate the template from synthtool?

blakeli0 commented on library_generation/repo-level-postprocess/apply_current_versions.sh:
```diff
@@ -0,0 +1,26 @@
+
+# This script sets the "current-version" written in versions.txt applied to all
+# pom.xml files in this monorepo.
+# This script plays supplemental role just in case Release Please pull request
```
Since the hermetic scripts read from `versions.txt`, do we still need this file? We may still need it for now, but I'm expecting poms to be generated correctly from `versions.txt` in the end.

JoeWang1127 replied:
I think this is a fallback function in case release please is not working correctly.

JoeWang1127 replied:
This script is removed.

blakeli0 commented on library_generation/new_client/templates/owlbot.py.j2:
```diff
@@ -0,0 +1,24 @@
+# Copyright 2021 Google LLC
```
This seems to be an existing bug. I see that we have modules generated [last year ](https://github.com/googleapis/google-cloud-java/blob/4982e78e416fec80baff99f0131815e2d4955acf/java-alloydb/owlbot.py#L1) with year `2021`. We need to at least update the copyright year to `2024` for all the templates for now, and come up with a way to always update it to current year.

JoeWang1127 replied:
Sure, I'll update the year here.

blakeli0 commented on library_generation/new_client/new-client.py:
```diff
@@ -0,0 +1,458 @@
+        client_input.include_samples,
+        "--versions_file",
+        f"{repo_root_dir}/versions.txt"],
+        cwd=repo_root_dir
```
After @diegomarquezp 's PR is in, the logic here needs to be updated to apply generate_library.sh and postprocess_lirbary separately.

blakeli0 commented on library_generation/new_client/new-client.py:
```diff
@@ -0,0 +1,458 @@
+        )
+
+    print("Applying the versions")
+    subprocess.check_call(
```
As I mentioned in another [comment](https://github.com/googleapis/sdk-platform-java/pull/2333/files#r1461254257), can we remove this script?

JoeWang1127 replied:
Removed

blakeli0 commented on library_generation/new_client/new-client.py:
```diff
@@ -0,0 +1,458 @@
+    distribution_name,
+    api_id,
+    requires_billing,
+    transport,
```
This should be removed in the future, the value of `transport` should be parsed from `Build.bazel`.

suztomo commented:
> `--generator-version=2.31.0`

I don't like release managers need to specify this value. Is there a way not to specify this argument?

JoeWang1127 commented on library_generation/new_client/new-client.py:
Do you mean the .gitignore in `library_generation/owlbot`?

JoeWang1127 replied:
I added an entry in `.gitignore` (in the repo root).

JoeWang1127 commented:
> I don't like release managers need to specify this value. Is there a way not to specify this argument?

One option is add a release-please annotation so that the version num is the same as the latest release version.

@suztomo what do you think this idea.

blakeli0 commented:
> > I don't like release managers need to specify this value. Is there a way not to specify this argument?
> 
> One option is add a release-please annotation so that the version num is the same as the latest release version.
> 
> @suztomo what do you think this idea.

Also I want to note that this is a temporary solution, we will add a new section for a new client library in the generation_config.yaml in the future. So specifying a generator version looks fine to me, and I would prefer us not to optimize too much for a temporary solution if it takes more than like 30 minutes.

suztomo commented:
I agree that Release Please is overkill for this work. Here is the shell script to get the version:

$ V=$(curl --silent 'https://raw.githubusercontent.com/googleapis/googleapis/master/WORKSPACE' |perl -nle 'print $1 if m/_gapic_generator_java_version\s+=\s+\"(.+)\"/')
$ echo $V
2.32.0

JoeWang1127 commented:
> I agree that Release Please is overkill for this work. Here is the shell script to get the version:
> 
> ```
> # Temporary solution to get gapic-generator-java version until the configuration file is ready
> $ V=$(curl --silent 'https://raw.githubusercontent.com/googleapis/googleapis/master/WORKSPACE' |perl -nle 'print $1 if m/_gapic_generator_java_version\s+=\s+\"(.+)\"/')
> $ echo $V
> 2.32.0
> ```

Thanks for the help.

JoeWang1127 commented on library_generation/new_client/new-client.py:
```diff
@@ -0,0 +1,445 @@
+        versioned_path=versioned_proto_path,
+    )
+    repo_root_dir = Path(f"{sys.path[0]}/../../").resolve()
+    generator_version = subprocess.check_output(
```
The version of generator is now parsed from WORKSPACE. This is a temporary fix so that release manager doesn't need to specify this value.

The plan is to use the version in config file later.

JoeWang1127 commented on library_generation/new_client/templates/owlbot.py.j2:
```diff
@@ -0,0 +1,28 @@
+
+s.remove_staging_dirs()
+{% if should_include_templates %}java.common_templates(monorepo=True, {% if template_excludes %}excludes=[
+{%- for exclude in template_excludes %}
```
Change template so that `update_owlbot_postprocessor_config.sh` is not needed.

JoeWang1127 commented on library_generation/owlbot/src/fix-poms.py:
```diff
@@ -477,52 +490,57 @@ def main(versions_file, monorepo):
             modules=modules,
             main_module=main_module,
+            monorepo=monorepo,
+            monorepo_version=monorepo_version
```
Change the template so that we can remove `set_parent_pom.sh`.

JoeWang1127 commented on library_generation/owlbot/src/fix-poms.py:
```diff
@@ -386,11 +399,12 @@ def main(versions_file, monorepo):
                 module=required_dependencies[path],
                 parent_module=parent_module,
                 main_module=main_module,
+                monorepo=monorepo,
```
Change the proto pom template to remove `consolidate_config.sh`.

suztomo reviewed (approved):
Looking forward to see the observations that it works for shopping CSS.

JoeWang1127 commented:
Steps to regenerate shopping CSS:
1. Checkout main branch of google-cloud-java and pull the latest change.
2. Copy `library_generation` to google-cloud-java's root directory.
3. Install python package:
```

```
4. Remove shopping CSS module in google-cloud-java:
```
rm -rf java-shopping-css
```
5. Run the generation command:
```
python library_generation/new_client/new-client.py generate --api_shortname="css" --proto-path="google/shopping/css" --name-pretty="CSS API" --product-docs="https://developers.google.com/comparison-shopping-services/api" --api-description="The CSS API is used to manage your CSS and control your CSS Products portfolio" --distribution-name="com.google.shopping:google-shopping-css" --destination-name=shopping-css
```
6. Verify the difference: [log](https://paste.googleplex.com/4844666342604800).
In addition, these files are removed because shopping has a types.proto that is not in the proto path (https://github.com/googleapis/sdk-platform-java/issues/2018):
```
deleted:    java-shopping-css/proto-google-shopping-css-v1/src/main/java/com/google/shopping/type/Channel.java
deleted:    java-shopping-css/proto-google-shopping-css-v1/src/main/java/com/google/shopping/type/ChannelOrBuilder.java
deleted:    java-shopping-css/proto-google-shopping-css-v1/src/main/java/com/google/shopping/type/CustomAttribute.java
deleted:    java-shopping-css/proto-google-shopping-css-v1/src/main/java/com/google/shopping/type/CustomAttributeOrBuilder.java
deleted:    java-shopping-css/proto-google-shopping-css-v1/src/main/java/com/google/shopping/type/Destination.java
deleted:    java-shopping-css/proto-google-shopping-css-v1/src/main/java/com/google/shopping/type/DestinationOrBuilder.java
deleted:    java-shopping-css/proto-google-shopping-css-v1/src/main/java/com/google/shopping/type/Price.java
deleted:    java-shopping-css/proto-google-shopping-css-v1/src/main/java/com/google/shopping/type/PriceOrBuilder.java
deleted:    java-shopping-css/proto-google-shopping-css-v1/src/main/java/com/google/shopping/type/ReportingContext.java
deleted:    java-shopping-css/proto-google-shopping-css-v1/src/main/java/com/google/shopping/type/ReportingContextOrBuilder.java
deleted:    java-shopping-css/proto-google-shopping-css-v1/src/main/java/com/google/shopping/type/TypesProto.java

```

suztomo commented:
> Verify the difference: [log](https://paste.googleplex.com/4844666342604800).

That looks good to me. Thank you.

blakeli0 commented on library_generation/owlbot/src/fix-poms.py:
```diff
@@ -477,52 +493,57 @@ def main(versions_file, monorepo):
             update_bom_pom(f"{artifact_id}-bom/pom.xml", modules)
     elif artifact_id+"-bom" not in excluded_poms_list:
         print("creating missing bom pom.xml")
+        monorepo_version = __get_monorepo_version(versions_file) \
```
I may missed this in earlier reviews, can you please remind me why we need to get `monorepo_version` now?

JoeWang1127 replied:
I want to generate the pom.xml in `*-bom` without using a repo-level post processing script (`set_parent_pom.sh` in this case) and run `mvn fmt:format` after the pom generation.

This requires to know the current version of `gapic-libraries-bom` and `google-cloud-jar-parent`, which is the same as the monorepo version.
